File: server-scala/src/main/scala/com/anontown/entities/res/Res.scala
package com.anontown.entities.res

import java.time.OffsetDateTime
import cats.implicits._
import com.anontown.AtError
import com.anontown.AtRightError
import com.anontown.AuthToken
import com.anontown.AtPrerequisiteError
import monocle.syntax.ApplyLens
import shapeless._
import record._
import com.anontown.utils.Record._
import com.anontown.entities.user.{UserId, User}
import com.anontown.entities.topic.TopicId
import cats.Applicative
import cats._, cats.implicits._, cats.derived._
import monocle.macros.syntax.lens.toGenApplyLensOps
import com.anontown.entities.topic.{
  TopicNormalId,
  TopicNormal,
  TopicForkId,
  TopicFork,
  Topic,
  TopicTemporaryId
}
import cats.data.EitherT
import com.anontown.ports.ConfigContainerAlg
import com.anontown.ports.ObjectIdGeneratorAlg
import com.anontown.ports.ClockAlg
import com.anontown.entities.history.{HistoryId, History}
import com.anontown.entities.topic.{TopicTemporary, UntaggedTopicId}
import com.anontown.entities.profile.{ProfileId, Profile}

sealed trait ResAPI {
  val id: String;
  val topicID: String;
  val date: String;
  val self: Option[Boolean];
  val uv: Int;
  val dv: Int;
  val hash: String;
  val replyCount: Int;
  val voteFlag: Option[VoteFlag];
}

object ResAPI {
  implicit val implEq: Eq[ResAPI] = {
    import auto.eq._
    semi.eq
  }
}

final case class ResForkAPI(
    id: String,
    topicID: String,
    date: String,
    self: Option[Boolean],
    uv: Int,
    dv: Int,
    hash: String,
    replyCount: Int,
    voteFlag: Option[VoteFlag],
    forkID: String
) extends ResAPI;

object ResForkAPI {
  implicit val implEq: Eq[ResForkAPI] = {
    import auto.eq._
    semi.eq
  }
}

final case class ResHistoryAPI(
    id: String,
    topicID: String,
    date: String,
    self: Option[Boolean],
    uv: Int,
    dv: Int,
    hash: String,
    replyCount: Int,
    voteFlag: Option[VoteFlag],
    historyID: String
) extends ResAPI;

object ResHistoryAPI {
  implicit val implEq: Eq[ResHistoryAPI] = {
    import auto.eq._
    semi.eq
  }
}

sealed trait ResNormalAPI extends ResAPI;

object ResNormalAPI {
  implicit val implEq: Eq[ResNormalAPI] = {
    import auto.eq._
    semi.eq
  }
}

final case class ResNormalActiveAPI(
    id: String,
    topicID: String,
    date: String,
    self: Option[Boolean],
    uv: Int,
    dv: Int,
    hash: String,
    replyCount: Int,
    voteFlag: Option[VoteFlag],
    name: Option[String],
    text: String,
    replyID: Option[String],
    profileID: Option[String],
    isReply: Option[Boolean]
) extends ResNormalAPI;

object ResNormalActiveAPI {
  implicit val implEq: Eq[ResNormalActiveAPI] = {
    import auto.eq._
    semi.eq
  }
}

final case class ResNormalDeleteAPI(
    id: String,
    topicID: String,
    date: String,
    self: Option[Boolean],
    uv: Int,
    dv: Int,
    hash: String,
    replyCount: Int,
    voteFlag: Option[VoteFlag],
    flag: ResDeleteReason
) extends ResNormalAPI;

object ResNormalDeleteAPI {
  implicit val implEq: Eq[ResNormalDeleteAPI] = {
    import auto.eq._
    semi.eq
  }
}

final case class ResTopicAPI(
    id: String,
    topicID: String,
    date: String,
    self: Option[Boolean],
    uv: Int,
    dv: Int,
    hash: String,
    replyCount: Int,
    voteFlag: Option[VoteFlag]
) extends ResAPI;

object ResTopicAPI {
  implicit val implEq: Eq[ResTopicAPI] = {
    import auto.eq._
    semi.eq
  }
}

sealed trait Res {
  type Self <: Res;
  type IdType <: TaggedResId;

  type TopicIdType <: TopicId;
  type ReplyResIdType <: ResId;

  type API <: ResAPI;

  type SelfApplyLens[T] = ApplyLens[Self, Self, T, T]

  def toAPI(
      authToken: Option[AuthToken]
  ): API;

  def id: IdType;
  def topic: TopicIdType;
  def date: OffsetDateTime;
  def user: UserId;
  def votesLens: SelfApplyLens[List[Vote]];
  def lv: Int;
  def hash: String;
  def replyCountLens: SelfApplyLens[Int];
}

object Res {
  implicit class ResService[A <: Res { type Self <: A }](val self: A) {
    type ResAPIIntrinsicProperty =
      ("id" ->> String) ::
        ("topicID" ->> String) ::
        ("date" ->> String) ::
        ("self" ->> Option[Boolean]) ::
        ("uv" ->> Int) ::
        ("dv" ->> Int) ::
        ("hash" ->> String) ::
        ("replyCount" ->> Int) ::
        ("voteFlag" ->> Option[VoteFlag]) ::
        HNil

    def resAPIIntrinsicProperty(
        authToken: Option[AuthToken]
    ): ResAPIIntrinsicProperty = {
      Record(
        id = self.id.value,
        topicID = self.topic.value,
        date = self.date.toString,
        self = authToken.map(_.user === self.user),
        uv = self.votesLens.get.filter(x => x.value > 0).size,
        dv = self.votesLens.get.filter(x => x.value < 0).size,
        hash = self.hash,
        replyCount = self.replyCountLens.get,
        voteFlag = authToken.map(
          authToken =>
            self.votesLens.get
              .find(authToken.user === _.user)
              .map(
                vote =>
                  if (vote.value > 0) VoteFlag.Uv()
                  else VoteFlag.Dv()
              )
              .getOrElse(VoteFlag.Not())
        )
      )
    }

    // 既に投票していたらエラー
    def vote(
        resUser: User,
        user: User,
        vtype: VoteType,
        authToken: AuthToken
    ): Either[AtError, (A, User)] = {
      type Result[A] = Either[AtError, A];

      assert(resUser.id === self.user);
      assert(user.id === authToken.user);

      for {
        _ <- Applicative[Result].whenA(user.id === self.user)(
          Left(new AtRightError("自分に投票は出来ません"))
        )
        _ <- Applicative[Result].whenA(
          self.votesLens.get.exists(_.user === user.id)
        )(Left(new AtPrerequisiteError("既に投票しています")))

        val valueAbs = (user.lv.toDouble / 100.0).floor.toInt + 1
        val value = vtype match {
          case VoteType.Uv() => valueAbs;
          case VoteType.Dv() => -valueAbs;
        }
        resUser <- Applicative[Result].pure(
          resUser.changeLv(resUser.lv + value)
        )
      } yield (
        self.votesLens.modify(
          _.appended(Vote(user = user.id, value = value))
        ),
        resUser
      )
    }

    def resetAndVote(
        resUser: User,
        user: User,
        vtype: VoteType,
        authToken: AuthToken
    ): Either[AtError, (A, User)] = {
      assert(resUser.id === self.user);
      assert(user.id === authToken.user);

      for {
        (res, resUser) <- self.votesLens.get
          .find(_.user === user.id)
          .filter(
            voted =>
              (
                voted.value > 0 && vtype === VoteType
                  .Uv()
              ) || (voted.value < 0 && vtype === VoteType
                .Dv())
          )
          .as(self.cv(resUser, user, authToken))
          .getOrElse(Right((self, resUser)))

        result <- res.vote(resUser, user, vtype, authToken)
      } yield result
    }

    def cv(
        resUser: User,
        user: User,
        authToken: AuthToken
    ): Either[AtError, (A, User)] = {
      type Result[A] = Either[AtError, A];

      assert(resUser.id === self.user);
      assert(user.id === authToken.user);

      for {
        vote <- self.votesLens.get
          .find(_.user === user.id)
          .liftTo(new AtPrerequisiteError("投票していません"))
        resUser <- Applicative[Result].pure(
          resUser.changeLv(resUser.lv - vote.value)
        )
      } yield (
        self.votesLens.modify(_.filter(_.user =!= user.id)),
        resUser
      )
    }
  }
}

final case class ResFork(
    id: ResForkId,
    topic: TopicNormalId,
    date: OffsetDateTime,
    user: UserId,
    votes: List[Vote],
    lv: Int,
    hash: String,
    replyCount: Int,
    fork: TopicForkId
) extends Res {
  override type Self = ResFork;
  override type IdType = ResForkId;
  override type TopicIdType = TopicNormalId;
  override type ReplyResIdType = Nothing;

  override type API = ResForkAPI

  override def votesLens = this.lens(_.votes)
  override def replyCountLens =
    this.lens(_.replyCount)

  override def toAPI(authToken: Option[AuthToken]): API = {
    LabelledGeneric[ResForkAPI].from(
      this
        .resAPIIntrinsicProperty(authToken)
        .merge(Record(forkID = this.fork.value))
    )
  }
}

object ResFork {
  def create[F[_]: Monad: ObjectIdGeneratorAlg: ClockAlg: ConfigContainerAlg](
      topic: TopicNormal,
      user: User,
      authToken: AuthToken,
      fork: TopicFork
  ): EitherT[
    F,
    AtError,
    (ResFork, TopicNormal)
  ] = {
    assert(user.id === authToken.user);
    for {
      requestDate <- EitherT.right(ClockAlg[F].getRequestDate())
      id <- EitherT.right(ObjectIdGeneratorAlg[F].generateObjectId())

      hash <- EitherT.right(topic.hash[F](user))

      val result = ResFork(
        fork = fork.id,
        id = ResForkId(id),
        topic = topic.id,
        date = requestDate,
        user = user.id,
        votes = List(),
        lv = user.lv * 5,
        hash = hash,
        replyCount = 0
      )
      topic <- EitherT.fromEither[F](topic.resUpdate(result, false))
    } yield (result, topic)
  }
}

final case class ResHistory(
    id: ResHistoryId,
    topic: TopicNormalId,
    date: OffsetDateTime,
    user: UserId,
    votes: List[Vote],
    lv: Int,
    hash: String,
    replyCount: Int,
    history: HistoryId
) extends Res {
  override type Self = ResHistory;
  override type IdType = ResHistoryId;
  override type TopicIdType = TopicNormalId;
  override type ReplyResIdType = Nothing;

  override type API = ResHistoryAPI

  override def votesLens = this.lens(_.votes)
  override def replyCountLens =
    this.lens(_.replyCount)

  override def toAPI(authToken: Option[AuthToken]): API = {
    LabelledGeneric[ResHistoryAPI].from(
      this
        .resAPIIntrinsicProperty(authToken)
        .merge(
          Record(
            historyID = this.history.value
          )
        )
    )
  }
}

object ResHistory {
  def create[F[_]: Monad: ObjectIdGeneratorAlg: ClockAlg: ConfigContainerAlg](
      topic: TopicNormal,
      user: User,
      authToken: AuthToken,
      history: History
  ): EitherT[
    F,
    AtError,
    (ResHistory, TopicNormal)
  ] = {
    assert(user.id === authToken.user);
    for {
      requestDate <- EitherT.right(ClockAlg[F].getRequestDate())
      id <- EitherT.right(
        ObjectIdGeneratorAlg[F].generateObjectId()
      )

      hash <- EitherT.right(topic.hash[F](user))

      val result = ResHistory(
        history = history.id,
        id = ResHistoryId(id),
        topic = topic.id,
        date = requestDate,
        user = user.id,
        votes = List(),
        lv = user.lv * 5,
        hash = hash,
        replyCount = 0
      )
      topic <- EitherT.fromEither[F](topic.resUpdate(result, false))
    } yield (result, topic)
  }
}

final case class ResNormal[TopicIdTypeArg <: TopicId, ReplyResId <: ResId](
    id: ResNormalId,
    topic: TopicIdTypeArg,
    date: OffsetDateTime,
    user: UserId,
    votes: List[Vote],
    lv: Int,
    hash: String,
    replyCount: Int,
    name: Option[ResName],
    text: ResText,
    reply: Option[Reply[ReplyResId]],
    deleteFlag: Option[ResDeleteReason],
    profile: Option[ProfileId],
    age: Boolean
) extends Res {
  override type Self = ResNormal[TopicIdTypeArg, ReplyResId];
  override type IdType = ResNormalId;
  override type ReplyResIdType = ReplyResId;

  override type TopicIdType = TopicIdTypeArg;

  override type API = ResNormalAPI;

  override def votesLens = this.lens(_.votes)
  override def replyCountLens =
    this.lens(_.replyCount)

  override def toAPI(authToken: Option[AuthToken]): API = {
    this.deleteFlag match {
      case None =>
        LabelledGeneric[ResNormalActiveAPI].from(
          this
            .resAPIIntrinsicProperty(authToken)
            .merge(
              Record(
                name = this.name.map(_.value),
                text = this.text.value,
                replyID = this.reply.map(_.res.value),
                profileID = this.profile.map(_.value),
                isReply = authToken.flatMap(
                  authToken => this.reply.map(authToken.user === _.user)
                )
              )
            )
        )
      case Some(deleteFlag) =>
        LabelledGeneric[ResNormalDeleteAPI].from(
          this
            .resAPIIntrinsicProperty(authToken)
            .merge(
              Record(
                flag = deleteFlag
              )
            )
        )
    }
  }

  def replyResIdWiden[A >: ReplyResId <: ResId]: ResNormal[TopicIdTypeArg, A] =
    this.copy()

  def topicIdWiden[A >: TopicIdTypeArg <: TopicId]: ResNormal[A, ReplyResId] =
    this.copy()
}

object ResNormal {
  def create[F[_]: Monad: ObjectIdGeneratorAlg: ClockAlg: ConfigContainerAlg, ResIdType <: ResId, ResType <: Res { type Self <: ResType; type IdType <: ResIdType; }, TopicType <: Topic { type Self <: TopicType; }](
      topic: TopicType,
      user: User,
      authUser: AuthToken,
      name: Option[String],
      text: String,
      reply: Option[ResType],
      profile: Option[Profile],
      age: Boolean
  ): EitherT[
    F,
    AtError,
    (ResNormal[topic.IdType, ResIdType], User, TopicType)
  ] = {
    assert(user.id === authUser.user);

    for {
      (name, text) <- EitherT
        .fromEither[F](
          (
            name
              .map(ResName.fromString(_).map(Some(_)))
              .getOrElse(Right(None))
              .toValidated,
            ResText.fromString(text).toValidated
          ).mapN((_, _)).toEither
        )
        .leftWiden[AtError]

      _ <- EitherT
        .fromEither[F](
          Either.cond(
            profile.map(_.user === user.id).getOrElse(true),
            (),
            new AtRightError("自分のプロフィールを指定して下さい。")
          )
        )
        .leftWiden[AtError]

      // もしリプ先があるかつ、トピックがリプ先と違えばエラー
      _ <- EitherT
        .fromEither[F](
          Either.cond(
            reply
              .map(
                reply =>
                  UntaggedTopicId
                    .fromTopicId(reply.topic) === UntaggedTopicId
                    .fromTopicId(topic.id)
              )
              .getOrElse(true),
            (),
            new AtPrerequisiteError("他のトピックのレスへのリプは出来ません")
          )
        )
        .leftWiden[AtError]

      requestDate <- EitherT.right(ClockAlg[F].getRequestDate())

      user <- user.changeLastRes[F]()

      id <- EitherT.right(
        ObjectIdGeneratorAlg[F].generateObjectId()
      )

      hash <- EitherT.right(topic.hash[F](user))

      val result = ResNormal(
        name = name,
        text = text,
        reply = reply.map(
          reply => Reply(res = reply.id, user = reply.user)
        ),
        deleteFlag = None,
        profile = profile.map(_.id),
        age = age,
        id = ResNormalId(id),
        topic = topic.id,
        date = requestDate,
        user = user.id,
        votes = List(),
        lv = user.lv * 5,
        hash = hash,
        replyCount = 0
      )
      topic <- EitherT.fromEither[F](topic.resUpdate(result, result.age))
    } yield (result.replyResIdWiden, user, topic)
  }

  implicit class ResNormalService[TopicIdType <: TopicId, ReplyResId <: ResId](
      val self: ResNormal[TopicIdType, ReplyResId]
  ) {
    def del(
        resUser: User,
        authToken: AuthToken
    ): Either[AtError, (ResNormal[TopicIdType, ReplyResId], User)] = {
      type Result[A] = Either[AtError, A]

      assert(resUser.id === authToken.user);
      for {
        _ <- Either.cond(
          authToken.user === self.user,
          (),
          new AtRightError("人の書き込み削除は出来ません")
        )

        _ <- Either.cond(
          self.deleteFlag.isEmpty,
          (),
          new AtPrerequisiteError("既に削除済みです")
        )

        resUser <- Applicative[Result].pure(resUser.changeLv(resUser.lv - 1))
      } yield (
        (
          self.copy(
            deleteFlag = Some(ResDeleteReason.Self())
          ),
          resUser
        )
      )
    }
  }
}

final case class ResTopic[TopicArg <: TopicTemporaryId](
    id: ResTopicId,
    topic: TopicArg,
    date: OffsetDateTime,
    user: UserId,
    votes: List[Vote],
    lv: Int,
    hash: String,
    replyCount: Int
) extends Res {
  override type Self = ResTopic[TopicArg]

  override type IdType = ResTopicId;

  override type TopicIdType = TopicArg
  override type ReplyResIdType = Nothing;

  override type API = ResTopicAPI

  override def votesLens = this.lens(_.votes)
  override def replyCountLens =
    this.lens(_.replyCount)

  override def toAPI(authToken: Option[AuthToken]): API = {
    LabelledGeneric[ResTopicAPI].from(
      this
        .resAPIIntrinsicProperty(authToken)
    )
  }
}

object ResTopic {
  def create[F[_]: Monad: ObjectIdGeneratorAlg: ClockAlg: ConfigContainerAlg, TopicTemporaryType <: TopicTemporary { type Self = TopicTemporaryType }](
      topic: TopicTemporaryType,
      user: User,
      authToken: AuthToken
  ): EitherT[
    F,
    AtError,
    (ResTopic[topic.IdType], TopicTemporaryType)
  ] = {
    assert(user.id === authToken.user);

    for {
      requestDate <- EitherT.right(ClockAlg[F].getRequestDate())
      id <- EitherT.right(
        ObjectIdGeneratorAlg[F].generateObjectId()
      )

      hash <- EitherT.right(topic.hash[F](user))

      val result = ResTopic(
        id = ResTopicId(id),
        topic = topic.id,
        date = requestDate,
        user = user.id,
        votes = List(),
        lv = user.lv * 5,
        hash = hash,
        replyCount = 0
      )
      topic <- EitherT.fromEither[F](topic.resUpdate(result, false))
    } yield (result, topic)
  }
}

File: server/packages/server/src/entities/res/res.ts
import { option } from "fp-ts";
import { Option } from "fp-ts/lib/Option";
import { pipe } from "fp-ts/lib/pipeable";
import * as Im from "immutable";
import {
  AtPrerequisiteError,
  AtRightError,
  paramsErrorMaker,
} from "../../at-error";
import { IAuthToken } from "../../auth";
import { Constant } from "../../constant";
import { IObjectIdGenerator } from "../../ports/index";
import { Copyable } from "../../utils";
import { applyMixins } from "../../utils";
import { History } from "../history";
import { Profile } from "../profile";
import { Topic, TopicFork, TopicNormal, TopicOne } from "../topic";
import { User } from "../user";

export interface IVote {
  readonly user: string;
  readonly value: number;
}

export type ResType = "normal" | "history" | "topic" | "fork";

export type ResAPIType = ResType | "delete";

/*
## 共通
  * self: 認証していなければnull。自分の書き込みかどうか
  * voteFlag: 認証していなければnull。投票状況
## normal
  * isReply: 認証していないもしくはリプ先がない時null。自分に対するリプライかどうか
## delete
  * normalかつdeleteFlagがactiveでない時
*/
export type IResAPI =
  | IResNormalAPI
  | IResHistoryAPI
  | IResTopicAPI
  | IResForkAPI
  | IResDeleteAPI;

export interface IResBaseAPI<T extends ResAPIType> {
  readonly id: string;
  readonly topicID: string;
  readonly date: Date;
  readonly self: boolean | null;
  readonly uv: number;
  readonly dv: number;
  readonly hash: string;
  readonly replyCount: number;
  readonly voteFlag: VoteFlag | null;
  readonly type: T;
}

export interface IResNormalAPI extends IResBaseAPI<"normal"> {
  readonly name: string | null;
  readonly text: string;
  readonly replyID: string | null;
  readonly profileID: string | null;
  readonly isReply: boolean | null;
}

export interface IResHistoryAPI extends IResBaseAPI<"history"> {
  readonly historyID: string;
}

export interface IResTopicAPI extends IResBaseAPI<"topic"> {}

export interface IResForkAPI extends IResBaseAPI<"fork"> {
  readonly forkID: string;
}

export interface IResDeleteAPI extends IResBaseAPI<"delete"> {
  readonly flag: "self" | "freeze";
}

export type VoteFlag = "uv" | "dv" | "not";
export type ResDeleteFlag = "active" | "self" | "freeze";
export interface IReply {
  readonly res: string;
  readonly user: string;
}

export abstract class ResBase<T extends ResType, C extends ResBase<T, C>> {
  abstract readonly id: string;
  abstract readonly topic: string;
  abstract readonly date: Date;
  abstract readonly user: string;
  abstract readonly votes: Im.List<IVote>;
  abstract readonly lv: number;
  abstract readonly hash: string;
  abstract readonly type: T;
  abstract readonly replyCount: number;

  abstract copy(partial: Partial<ResBase<T, C>>): C;

  // TODO: 名前をresetAndVoteに
  v(
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ): { res: C; resUser: User } {
    const voted = this.votes.find(x => x.user === user.id);
    const data =
      voted !== undefined &&
      ((voted.value > 0 && type === "dv") || (voted.value < 0 && type === "uv"))
        ? this.cv(resUser, user, _authToken)
        : { res: this, resUser };
    return data.res._v(data.resUser, user, type, _authToken);
  }

  // TODO: 名前をvoteに
  _v(
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ): { res: C; resUser: User } {
    if (user.id === this.user) {
      throw new AtRightError("自分に投票は出来ません");
    }
    if (this.votes.find(x => x.user === user.id) !== undefined) {
      throw new AtPrerequisiteError("既に投票しています");
    }
    const valueAbs = Math.floor(user.lv / 100) + 1;
    const value = type === "uv" ? valueAbs : -valueAbs;
    const newResUser = resUser.changeLv(resUser.lv + value);
    return {
      res: this.copy({
        votes: this.votes.push({ user: user.id, value }),
      }),
      resUser: newResUser,
    };
  }

  cv(
    resUser: User,
    user: User,
    _authToken: IAuthToken,
  ): { res: C; resUser: User } {
    const vote = this.votes.find(x => x.user === user.id);
    if (vote === undefined) {
      throw new AtPrerequisiteError("投票していません");
    }
    const newResUser = resUser.changeLv(resUser.lv - vote.value);
    return {
      res: this.copy({
        votes: this.votes.remove(this.votes.indexOf(vote)),
      }),
      resUser: newResUser,
    };
  }

  toBaseAPI(authToken: Option<IAuthToken>): IResBaseAPI<T> {
    const voteFlag = pipe(
      authToken,
      option.map(authToken => {
        const vote = this.votes.find(v => authToken.user === v.user);
        if (vote === undefined) {
          return "not";
        } else {
          return vote.value > 0 ? "uv" : "dv";
        }
      }),
      option.toNullable,
    );

    return {
      id: this.id,
      topicID: this.topic,
      date: this.date,
      self: pipe(
        authToken,
        option.map(authToken => authToken.user === this.user),
        option.toNullable,
      ),
      uv: this.votes.filter(x => x.value > 0).size,
      dv: this.votes.filter(x => x.value < 0).size,
      hash: this.hash,
      replyCount: this.replyCount,
      voteFlag,
      type: this.type,
    };
  }
}

export type Res = ResNormal | ResHistory | ResTopic | ResFork;

export class ResNormal extends Copyable<ResNormal>
  implements ResBase<"normal", ResNormal> {
  static create(
    objidGenerator: IObjectIdGenerator,
    topic: Topic,
    user: User,
    _authToken: IAuthToken,
    name: Option<string>,
    text: string,
    reply: Option<Res>,
    profile: Option<Profile>,
    age: boolean,
    now: Date,
  ) {
    paramsErrorMaker([
      {
        field: "text",
        val: text,
        regex: Constant.res.text.regex,
        message: Constant.res.text.msg,
      },
      {
        field: "name",
        val: name,
        regex: Constant.res.name.regex,
        message: Constant.res.name.msg,
      },
    ]);

    if (
      pipe(
        profile,
        option.map(profile => profile.user !== user.id),
        option.getOrElse(() => false),
      )
    ) {
      throw new AtRightError("自分のプロフィールを指定して下さい。");
    }

    // もしリプ先があるかつ、トピックがリプ先と違えばエラー
    if (
      pipe(
        reply,
        option.map(reply => reply.topic !== topic.id),
        option.getOrElse(() => false),
      )
    ) {
      throw new AtPrerequisiteError("他のトピックのレスへのリプは出来ません");
    }

    const newUser = user.changeLastRes(now);

    const result = new ResNormal(
      name,
      text,
      pipe(
        reply,
        option.map(reply => ({ res: reply.id, user: reply.user })),
      ),
      "active",
      pipe(
        profile,
        option.map(profile => profile.id),
      ),
      age,
      objidGenerator.generateObjectId(),
      topic.id,
      now,
      newUser.id,
      Im.List(),
      newUser.lv * 5,
      topic.hash(now, newUser),
      0,
    );

    const newTopic = topic.resUpdate(result);
    return { res: result, user: newUser, topic: newTopic };
  }

  readonly type: "normal" = "normal";

  toBaseAPI!: (authToken: Option<IAuthToken>) => IResBaseAPI<"normal">;
  cv!: (
    resUser: User,
    user: User,
    _authToken: IAuthToken,
  ) => { res: ResNormal; resUser: User };
  _v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResNormal; resUser: User };
  v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResNormal; resUser: User };

  constructor(
    readonly name: Option<string>,
    readonly text: string,
    readonly reply: Option<IReply>,
    readonly deleteFlag: ResDeleteFlag,
    readonly profile: Option<string>,
    readonly age: boolean,
    readonly id: string,
    readonly topic: string,
    readonly date: Date,
    readonly user: string,
    readonly votes: Im.List<IVote>,
    readonly lv: number,
    readonly hash: string,
    readonly replyCount: number,
  ) {
    super(ResNormal);
  }

  toAPI(authToken: Option<IAuthToken>): IResNormalAPI | IResDeleteAPI {
    if (this.deleteFlag === "active") {
      return {
        ...this.toBaseAPI(authToken),
        name: pipe(this.name, option.toNullable),
        text: this.text,
        replyID: pipe(
          this.reply,
          option.map(x => x.res),
          option.toNullable,
        ),
        profileID: pipe(this.profile, option.toNullable),
        isReply: pipe(
          authToken,
          option.chain(authToken =>
            pipe(
              this.reply,
              option.map(reply => authToken.user === reply.user),
            ),
          ),
          option.toNullable,
        ),
      };
    } else {
      return {
        ...this.toBaseAPI(authToken),
        type: "delete",
        flag: this.deleteFlag,
      };
    }
  }

  del(resUser: User, authToken: IAuthToken) {
    if (authToken.user !== this.user) {
      throw new AtRightError("人の書き込み削除は出来ません");
    }

    if (this.deleteFlag !== "active") {
      throw new AtPrerequisiteError("既に削除済みです");
    }

    const newResUser = resUser.changeLv(resUser.lv - 1);
    return {
      res: this.copy({
        deleteFlag: "self",
      }),
      resUser: newResUser,
    };
  }
}
applyMixins(ResNormal, [ResBase]);

export class ResHistory extends Copyable<ResHistory>
  implements ResBase<"history", ResHistory> {
  static create(
    objidGenerator: IObjectIdGenerator,
    topic: TopicNormal,
    user: User,
    _authToken: IAuthToken,
    history: History,
    now: Date,
  ) {
    const result = new ResHistory(
      history.id,
      objidGenerator.generateObjectId(),
      topic.id,
      now,
      user.id,
      Im.List(),
      user.lv * 5,
      topic.hash(now, user),
      0,
    );

    const newTopic = topic.resUpdate(result);
    return { res: result, topic: newTopic };
  }

  toBaseAPI!: (authToken: Option<IAuthToken>) => IResBaseAPI<"history">;
  cv!: (
    resUser: User,
    user: User,
    _authToken: IAuthToken,
  ) => { res: ResHistory; resUser: User };
  _v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResHistory; resUser: User };
  v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResHistory; resUser: User };

  readonly type: "history" = "history";

  constructor(
    readonly history: string,
    readonly id: string,
    readonly topic: string,
    readonly date: Date,
    readonly user: string,
    readonly votes: Im.List<IVote>,
    readonly lv: number,
    readonly hash: string,
    readonly replyCount: number,
  ) {
    super(ResHistory);
  }

  toAPI(authToken: Option<IAuthToken>): IResHistoryAPI {
    return {
      ...this.toBaseAPI(authToken),
      historyID: this.history,
    };
  }
}
applyMixins(ResHistory, [ResBase]);

export class ResTopic extends Copyable<ResTopic>
  implements ResBase<"topic", ResTopic> {
  static create<TC extends TopicOne | TopicFork>(
    objidGenerator: IObjectIdGenerator,
    topic: TC,
    user: User,
    _authToken: IAuthToken,
    now: Date,
  ) {
    const result = new ResTopic(
      objidGenerator.generateObjectId(),
      topic.id,
      now,
      user.id,
      Im.List(),
      user.lv * 5,
      topic.hash(now, user),
      0,
    );

    // TODO:キャストなしで書きたい
    const newTopic = topic.resUpdate(result) as TC;
    return { res: result, topic: newTopic };
  }

  toBaseAPI!: (authToken: Option<IAuthToken>) => IResBaseAPI<"topic">;
  cv!: (
    resUser: User,
    user: User,
    _authToken: IAuthToken,
  ) => { res: ResTopic; resUser: User };
  _v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResTopic; resUser: User };
  v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResTopic; resUser: User };

  readonly type: "topic" = "topic";

  constructor(
    readonly id: string,
    readonly topic: string,
    readonly date: Date,
    readonly user: string,
    readonly votes: Im.List<IVote>,
    readonly lv: number,
    readonly hash: string,
    readonly replyCount: number,
  ) {
    super(ResTopic);
  }

  toAPI(authToken: Option<IAuthToken>): IResTopicAPI {
    return this.toBaseAPI(authToken);
  }
}
applyMixins(ResTopic, [ResBase]);

export class ResFork extends Copyable<ResFork>
  implements ResBase<"fork", ResFork> {
  static create(
    objidGenerator: IObjectIdGenerator,
    topic: TopicNormal,
    user: User,
    _authToken: IAuthToken,
    fork: TopicFork,
    now: Date,
  ) {
    const result = new ResFork(
      fork.id,
      objidGenerator.generateObjectId(),
      topic.id,
      now,
      user.id,
      Im.List(),
      user.lv * 5,
      topic.hash(now, user),
      0,
    );

    const newTopic = topic.resUpdate(result);
    return { res: result, topic: newTopic };
  }

  toBaseAPI!: (authToken: Option<IAuthToken>) => IResBaseAPI<"fork">;
  cv!: (
    resUser: User,
    user: User,
    _authToken: IAuthToken,
  ) => { res: ResFork; resUser: User };
  _v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResFork; resUser: User };
  v!: (
    resUser: User,
    user: User,
    type: "uv" | "dv",
    _authToken: IAuthToken,
  ) => { res: ResFork; resUser: User };

  readonly type: "fork" = "fork";

  constructor(
    readonly fork: string,
    readonly id: string,
    readonly topic: string,
    readonly date: Date,
    readonly user: string,
    readonly votes: Im.List<IVote>,
    readonly lv: number,
    readonly hash: string,
    readonly replyCount: number,
  ) {
    super(ResFork);
  }

  toAPI(authToken: Option<IAuthToken>): IResForkAPI {
    return {
      ...this.toBaseAPI(authToken),
      forkID: this.fork,
    };
  }
}
applyMixins(ResFork, [ResBase]);
